include/revng/Model/CommonTypeMethods.h

```diff
@@ -0,0 +1,224 @@
+
+  /// This is a specialized version of \ref asDefinition(), see
+  /// the documentation there for more details.
+  model::StructDefinition &asStruct() { return *tryGetAsStruct(); }
+  const model::StructDefinition &asStruct() const { return *tryGetAsStruct(); }
+  model::StructDefinition *tryGetAsStruct();
+  const model::StructDefinition *tryGetAsStruct() const;
+  bool isStruct() const { return tryGetAsStruct() != nullptr; }
+
+  /// This is a specialized version of \ref asDefinition(), see
+  /// the documentation there for more details.
+  model::UnionDefinition &asUnion() { return *tryGetAsUnion(); }
+  const model::UnionDefinition &asUnion() const { return *tryGetAsUnion(); }
+  model::UnionDefinition *tryGetAsUnion();
+  const model::UnionDefinition *tryGetAsUnion() const;
+  bool isUnion() const { return tryGetAsUnion() != nullptr; }
+
+  /// This is a specialized version of \ref asDefinition(), see
+  /// the documentation there for more details.
+  model::EnumDefinition &asEnum() { return *tryGetAsEnum(); }
+  const model::EnumDefinition &asEnum() const { return *tryGetAsEnum(); }
+  model::EnumDefinition *tryGetAsEnum();
+  const model::EnumDefinition *tryGetAsEnum() const;
+  bool isEnum() const { return tryGetAsEnum() != nullptr; }
+
+  /// This is a specialized version of \ref asDefinition(), see
+  /// the documentation there for more details.
+  model::PrimitiveType &asPrimitive() { return *tryGetAsPrimitive(); }
+  const auto &asPrimitive() const { return *tryGetAsPrimitive(); }
+  model::PrimitiveType *tryGetAsPrimitive();
+  const model::PrimitiveType *tryGetAsPrimitive() const;
+  bool isPrimitive() const { return tryGetAsPrimitive() != nullptr; }
+
+  /// This is a specialized version of \ref asDefinition(), see
+  /// the documentation there for more details.
+  model::ArrayType &asArray() { return *tryGetAsArray(); }
+  const model::ArrayType &asArray() const { return *tryGetAsArray(); }
+  model::ArrayType *tryGetAsArray();
+  const model::ArrayType *tryGetAsArray() const;
+  bool isArray() const { return tryGetAsArray() != nullptr; }
+
+  /// This is a specialized version of \ref asDefinition(), see
+  /// the documentation there for more details.
+  model::PointerType &asPointer() { return *tryGetAsPointer(); }
+  const model::PointerType &asPointer() const { return *tryGetAsPointer(); }
+  model::PointerType *tryGetAsPointer();
+  const model::PointerType *tryGetAsPointer() const;
+  bool isPointer() const { return tryGetAsPointer() != nullptr; }
+
+  /// This helper returns true if and only if this type is is a primitive or
+  /// a pointer.
+  ///
+  /// \note it asserts on `void`.
+  bool isScalar() const;
+
+public:
+  /// Use this method to check whether a specific type is constant or not, as
+  /// opposed to the \ref IsConst() member (which should be reserved for
+  /// assigning `const`ness), since it unwraps typedefs and arrays.
+  bool isConst() const;
```

Comment:
Let's consider methods returning `bool` (such as `isConst`, `isScalar`, `isPointer`...).
I think we need to clearly distinguish functions that make this check locally and those that recur over `typedef`s.

I'm fine with having `isScalar` being recursive, but I'd like to 1) have the docs state that's going to do that, 2) have another method `isLocallyScalar` that does not recur over `typedef`s and it's used to implement `isScalar`.
Alternative naming would be `isScalar` for local check and `isOrTypedefsScalar`/`isRecursivelyScalar` for the recursive one.

Clearly, `isLocallyConst` does not make sense because that's simply `IsConst()` but for the others I think it makes sense.

Reply:
I think we have a clearer way forward on this matter. Shall we mark this as resolved?